tests/data/semparse/type_declarations/type_declaration_test.py

```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
         type_a = types.NamedBasicType("A")
         type_b = types.NamedBasicType("B")
         assert type_a.resolve(type_b) is None
+
+    def test_valid_actions(self):
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <e,<r,<d,r>>>
+        type_signatures = {'F': ComplexType(type_e, ComplexType(type_r, ComplexType(type_d, type_r)))}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<e,<r,<d,r>>>"] == {"<e,<r,<d,r>>> -> sample_function"}
+        assert valid_actions["<r,<d,r>>"] == {"<r,<d,r>> -> [<e,<r,<d,r>>>, e]"}
+        assert valid_actions["<d,r>"] == {"<d,r> -> [<r,<d,r>>, r]"}
+        assert valid_actions["r"] == {"r -> [<d,r>, d]"}
+
+    def test_valid_actions_with_placeholder_type(self):
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <#1,#1>
+        type_signatures = {'F': types.IdentityType(types.ANY_TYPE, types.ANY_TYPE)}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<#1,#1>"] == {"<#1,#1> -> sample_function"}
```

Comment:
Does it make sense to add rules like `<e,e> -> sample_function`, and `e -> [<e,e>, e]`?  Or will leaving the placeholders in always be sufficient?

Reply:
Practically, leaving them as placeholders should always be sufficient. The action sequences contain placeholder strings in terminal productions, and they appear as resolved types after they are applied to other types: https://github.com/allenai/allennlp/blob/wikitables/tests/data/semparse/worlds/wikitables_world_test.py#L32-L37